libraries/standard/mqtt/src/iot_mqtt_network.c

```diff
@@ -233,12 +246,61 @@ static bool _incomingPacketValid( uint8_t packetType )
 
 /*-----------------------------------------------------------*/
 
+static IotMqttError_t _allocateAndReceivePacket( IotNetworkConnection_t pNetworkConnection,
+                                                 const _mqttConnection_t * pMqttConnection,
+                                                 _mqttPacket_t * pIncomingPacket )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+    size_t dataBytesRead = 0;
+
+    IotMqtt_Assert( pMqttConnection != NULL );
+    IotMqtt_Assert( pNetworkConnection != NULL );
+    IotMqtt_Assert( pIncomingPacket != NULL );
+
+    /* Allocate a buffer for the remaining data and read the data. */
+    if( pIncomingPacket->remainingLength > 0U )
+    {
+        pIncomingPacket->pRemainingData = IotMqtt_MallocMessage( pIncomingPacket->remainingLength );
+
+        if( pIncomingPacket->pRemainingData == NULL )
+        {
+            /* In some implementations IotLogError() maps to C standard printing API
+             * that need specific primitive types for format specifiers. Also,
+             * inttypes.h may not be available on some C99 compilers, despite stdint.h
+             * being available. */
+            /* coverity[misra_c_2012_directive_4_6_violation] */
+            IotLogError( "(MQTT connection %p) Failed to allocate buffer of length "
+                         "%lu for incoming packet type %lu.",
+                         pMqttConnection,
+                         ( unsigned long ) pIncomingPacket->remainingLength,
+                         ( unsigned long ) pIncomingPacket->type );
+
+            _flushPacket( pNetworkConnection, pMqttConnection, pIncomingPacket->remainingLength );
+
+            status = IOT_MQTT_NO_MEMORY;
+        }
+
+        if( status == IOT_MQTT_SUCCESS )
+        {
+            dataBytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
+                                                                         pIncomingPacket->pRemainingData,
+                                                                         pIncomingPacket->remainingLength );
+
+            if( dataBytesRead != pIncomingPacket->remainingLength )
+            {
+                status = IOT_MQTT_BAD_RESPONSE;
```

Comment:
Should we add a logging statement here, or is the bad response status code enough?